lib/srv/discovery/fetchers/db/aws.go

```diff
@@ -17,11 +17,122 @@ limitations under the License.
+}
+
+// awsFetcherConfig is the AWS database fetcher configuration.
+type awsFetcherConfig struct {
+	// AWSClients are the AWS API clients.
+	AWSClients cloud.AWSClients
+	// Type is the type of DB matcher, for example "rds", "redshift", etc.
+	Type string
+	// AssumeRole provides a role ARN and ExternalID to assume an AWS role
+	// when fetching databases.
+	AssumeRole types.AssumeRole
+	// Labels is a selector to match cloud database tags.
+	Labels types.Labels
+	// Region is the AWS region selector to match cloud databases.
+	Region string
+	// Log is a field logger to provide structured logging for each matcher,
+	// based on its config settings by default.
+	Log logrus.FieldLogger
+}
+
+// CheckAndSetDefaults validates the config and sets defaults.
+func (cfg *awsFetcherConfig) CheckAndSetDefaults(component string) error {
+	if cfg.AWSClients == nil {
+		return trace.BadParameter("missing parameter AWSClients")
+	}
+	if cfg.Type == "" {
+		return trace.BadParameter("missing parameter Type")
+	}
+	if len(cfg.Labels) == 0 {
+		return trace.BadParameter("missing parameter Labels")
+	}
+	if cfg.Region == "" {
+		return trace.BadParameter("missing parameter Region")
+	}
+	if cfg.Log == nil {
+		cfg.Log = logrus.WithFields(logrus.Fields{
+			trace.Component: "watch:" + component,
+			"labels":        cfg.Labels,
+			"region":        cfg.Region,
+			"role":          cfg.AssumeRole,
+		})
+	}
+	return nil
+}
+
+// newAWSFetcher returns a AWS database fetcher for the provided selectors
+// and AWS database-type specific fetcher plugin.
+func newAWSFetcher(cfg awsFetcherConfig, plugin awsFetcherPlugin) (*awsFetcher, error) {
+	if err := cfg.CheckAndSetDefaults(plugin.ComponentShortName()); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &awsFetcher{cfg: cfg, plugin: plugin}, nil
+}
+
 // awsFetcher is the common base for AWS database fetchers.
 type awsFetcher struct {
+	// cfg is the awsFetcher configuration.
+	cfg awsFetcherConfig
+	// plugin does AWS database type specific API calls fetch databases.
+	plugin awsFetcherPlugin
+}
+
+// awsFetcher implements common.Fetcher.
+var _ common.Fetcher = (*awsFetcher)(nil)
+
+// Get returns AWS databases matching the fetcher's selectors.
+func (f *awsFetcher) Get(ctx context.Context) (types.ResourcesWithLabels, error) {
+	databases, err := f.getDatabases(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	f.rewriteDatabases(databases)
+	return databases.AsResources(), nil
+}
+
+func (f *awsFetcher) getDatabases(ctx context.Context) (types.Databases, error) {
+	databases, err := f.plugin.GetDatabases(ctx, &f.cfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return filterDatabasesByLabels(databases, f.cfg.Labels, f.cfg.Log), nil
+}
+
+// rewriteDatabases rewrites the discovered databases.
+func (f *awsFetcher) rewriteDatabases(databases types.Databases) {
+	for _, db := range databases {
+		f.applyAssumeRole(db)
+		common.ApplyAWSDatabaseNameSuffix(db, f.cfg.Type)
```

Comment:
nit: wondering if the renaming can happen within `serverices.NewDatabaseFromXXXX`

Reply:
I considered doing that, but I was worried I would miss one, or try to build the name before all of the relevant metadata was collected to build the suffix, or we'd forget to do it in a future change. That was why I refactored the aws fetchers into plugins, so they all go through this path at the end just before the database gets sent to the reconciler

Reply:
sounds good